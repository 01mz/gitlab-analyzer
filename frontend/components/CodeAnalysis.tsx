import React, {useEffect} from "react";
import { makeStyles, withStyles, Theme, createStyles } from '@material-ui/core/styles';
import { green, purple } from '@material-ui/core/colors';
import FormGroup from '@material-ui/core/FormGroup';
import FormControlLabel from '@material-ui/core/FormControlLabel';
import Checkbox, { CheckboxProps } from '@material-ui/core/Checkbox';
import Avatar from '@material-ui/core/Avatar';
import axios, {AxiosResponse} from "axios";
import {useRouter} from "next/router";

const GreenCheckbox = withStyles({
    root: {
        color: green[400],
        '&$checked': {
            color: green[600],
        },
    },
    checked: {},
})((props: CheckboxProps) => <Checkbox color="default" {...props} />);

const PurpleCheckbox = withStyles({
    root: {
        color: purple[400],
        '&$checked': {
            color: purple[600],
        },
    },
    checked: {},
})((props: CheckboxProps) => <Checkbox color="default" {...props} />);

// TODO: add graph tag where placeholder is

const useStyles = makeStyles((theme: Theme) =>
    createStyles({
        container1: {
            display: 'flex',
            width: '100%',
        },
        container2: {
            display: 'flex',
            justifyContent: 'flex-start',
            width: '70%',
        },
        outerContainer: {
            flexDirection: 'column',
            width: '100%',
        },
        textContainer1: {
            flexDirection: 'column',
            margin: '0px 0px 0px 20px',
            padding: '0px',
        },
        textContainer2: {
            flexDirection: 'column',
            width:'30%',
        },
        repoNameText: {
            fontSize: '2em',
            margin: '16px 0px',
        },
        smallTextColor: {
            color: 'grey',
            margin: '0px 0px',
        },
        mrScoreText: {
            fontSize:'1.2em',
            margin: '16px 0px',
            textAlign: 'right'
        },
        commitScoreText: {
            fontSize:'1.2em',
            margin: '0px 0px',
            textAlign: 'right'
        },
        container3: {
            display: 'flex',
            justifyContent: 'space-around',
            margin: '24px 0px',
        },
        graphContainer: {
            display: 'flex',
            justifyContent: 'space-around',
            flexDirection: 'column',
        },
        avatarSize: {
            width: theme.spacing(15),
            height: theme.spacing(15),
        },
    }),
);

const CodeAnalysis = () => {
    const classes = useStyles();
    const [checkboxState, setCheckboxState] = React.useState({
        checkedCommitForGraphA: true,
        checkedMergeRequestForGraphA: true,
        checkedCommitForGraphB: true,
        checkedMergeRequestForGraphB: true,
    });

    const [mergerRequestCount, setMergerRequestCount] = React.useState<number>();
    const [projectName, setProjectName] = React.useState<String>();
    const [commitCount, setCommitCount] = React.useState<number>();
    const [mergeRequestScore, setMergeRequestScore] = React.useState<number>();
    const [commitScore, setCommitScore] = React.useState<number>();

    const router = useRouter();
    const { projectId, startDateTime, endDateTime } =  router.query;

    useEffect(() => {
        if (router.isReady) {
            axios
                .get(`${process.env.NEXT_PUBLIC_API_URL}/gitlab/projects/${projectId}`)
                .then((resp: AxiosResponse) => {
                    setProjectName(resp.data.name_with_namespace);
                });
            axios
                .get(`${process.env.NEXT_PUBLIC_API_URL}/gitlab/projects/${projectId}/merge_requests?startDateTime=${startDateTime}&endDateTime=${endDateTime}`)
                .then((resp: AxiosResponse) => {
                    setMergerRequestCount(resp.data.length);
                });
            axios
                .get(`${process.env.NEXT_PUBLIC_API_URL}/gitlab/projects/${projectId}/commits?startDateTime=${startDateTime}&endDateTime=${endDateTime}`)
                .then((resp: AxiosResponse) => {
                    setCommitCount(resp.data.length);
                });
            axios
                .get(`${process.env.NEXT_PUBLIC_API_URL}/data/projects/${projectId}/merge_requests/score?startDateTime=${startDateTime}&endDateTime=${endDateTime}`)
                .then((resp: AxiosResponse) => {
                    setMergeRequestScore(resp.data);
                });
            axios
                .get(`${process.env.NEXT_PUBLIC_API_URL}/data/projects/${projectId}/commits/score?startDateTime=${startDateTime}&endDateTime=${endDateTime}`)
                .then((resp: AxiosResponse) => {
                    setCommitScore(resp.data);
                });
        }
    }, [projectId]);

    const handleChange = (event: React.ChangeEvent<HTMLInputElement>) => {
        setCheckboxState({ ...checkboxState, [event.target.name]: event.target.checked });
    };

    return (
    <>
        <div className={classes.outerContainer}>
            <div className={classes.container1}>
                <div className={classes.container2}>
                    <Avatar className={classes.avatarSize} variant='square'>R</Avatar>
                    <div className={classes.textContainer1}>
                        <h1 className={classes.repoNameText}>{projectName}</h1>
                        <p className={classes.smallTextColor}>- {commitCount} Commits - {mergerRequestCount} Merge Request -</p>
                    </div>
                </div>
                <div className={classes.textContainer2}>
                    <p className={classes.mrScoreText}>Merge Request Score: {mergeRequestScore}</p>
                    <p className={classes.commitScoreText}>Commit Score: {commitScore}</p>
                </div>
            </div>
            <div className={classes.container3}>
                <div className={classes.graphContainer}>
                    <p> (graph placeholder)            </p>
                </div>
                <FormGroup>
                    <FormControlLabel
                        control={<GreenCheckbox checked={checkboxState.checkedCommitForGraphA} onChange={handleChange} name="checkedCommitForGraphA"/>}
                        label="Commits"
                    />
                    <FormControlLabel
                        control={<PurpleCheckbox checked={checkboxState.checkedMergeRequestForGraphA} onChange={handleChange} name="checkedMergeRequestForGraphA"/>}
                        label="Merge Requests"
                    />
                </FormGroup>
            </div>
            <div className={classes.container3}>
                <div className={classes.graphContainer}>
                    <p> (graph placeholder)            </p>
                </div>
                <FormGroup>
                    <FormControlLabel
                        control={<GreenCheckbox checked={checkboxState.checkedCommitForGraphB} onChange={handleChange} name="checkedCommitForGraphB"/>}
                        label="Commits"
                    />
                    <FormControlLabel
                        control={<PurpleCheckbox checked={checkboxState.checkedMergeRequestForGraphB} onChange={handleChange} name="checkedMergeRequestForGraphB"/>}
                        label="Merge Requests"
                    />
                </FormGroup>
            </div>
        </div>
    </>
    );
};

export default CodeAnalysis;