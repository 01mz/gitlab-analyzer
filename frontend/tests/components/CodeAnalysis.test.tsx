import React from 'react';
import {render} from '@testing-library/react';
import CodeAnalysis from '../../components/CodeAnalysis';

describe("Code Analysis", () =>{
    const useRouter = jest.spyOn(require('next/router'), 'useRouter');
    const mockUseEffect = jest.spyOn(React, 'useEffect')
    const mockAxios = jest.spyOn(require('axios'), 'get');

    beforeEach(()=>{
        useRouter.mockClear();
        mockUseEffect.mockClear();
        mockAxios.mockClear();
    })

    it("Snapshot serverId", () => {
        useRouter.mockImplementationOnce(() => ({
            query: { projectId: 'TestId', startDateTime: '2020-08-22T15:40-05:00', endDateTime:'2021-08-22T15:40-05:00' },
        }));
        const { container } = render(
            <CodeAnalysis />
        )
        expect(container).toMatchSnapshot();

    })
    it("Test useEffect", ()=>{
        useRouter.mockImplementationOnce(() => ({
            query: { projectId: 'TestId', startDateTime: '2020-08-22T15:40-05:00', endDateTime:'2021-08-22T15:40-05:00' },
        }));
        render(<CodeAnalysis />);
        expect(mockUseEffect).toBeCalled();

    })

    it('Test axios',()=>{
        useRouter.mockImplementationOnce(() => ({
            query: { projectId: 'TestId', startDateTime: '2020-08-22T15:40-05:00', endDateTime:'2021-08-22T15:40-05:00' },
            isReady: true,
        }));
        render(<CodeAnalysis />);
        expect(mockAxios).toHaveBeenCalledTimes(5);

        useRouter.mockImplementationOnce(() => ({
            query: { projectId: 'TestId', startDateTime: '2020-08-22T15:40-05:00', endDateTime:'2021-08-22T15:40-05:00' },
            isReady: false,
        }));
        render(<CodeAnalysis />);
        expect(mockAxios).toHaveBeenCalledTimes(5);
    })


})