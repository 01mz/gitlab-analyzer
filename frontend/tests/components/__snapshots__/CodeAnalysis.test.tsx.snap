// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Code Analysis Snapshot serverId 1`] = `
<div>
  <div
    class="makeStyles-outerContainer-3"
  >
    <div
      class="makeStyles-container1-1"
    >
      <div
        class="makeStyles-container2-2"
      >
        <div
          class="MuiAvatar-root MuiAvatar-square makeStyles-avatarSize-12 MuiAvatar-colorDefault"
        >
          R
        </div>
        <div
          class="makeStyles-textContainer1-4"
        >
          <h1
            class="makeStyles-repoNameText-6"
          >
            loading...
          </h1>
          <p
            class="makeStyles-smallTextColor-7"
          >
            - 
            loading...
             Commits - 
            loading...
             Merge Request -
          </p>
        </div>
      </div>
      <div
        class="makeStyles-textContainer2-5"
      >
        <p
          class="makeStyles-mrScoreText-8"
        >
          Merge Request Score: 
          loading...
        </p>
        <p
          class="makeStyles-commitScoreText-9"
        >
          Commit Score: 
          loading...
        </p>
      </div>
    </div>
    <div
      class="makeStyles-container3-10"
    >
      <div
        class="makeStyles-graphContainer-11"
      >
        <p>
           (graph placeholder)            
        </p>
      </div>
      <div
        class="MuiFormGroup-root"
      >
        <label
          class="MuiFormControlLabel-root"
        >
          <span
            aria-disabled="false"
            class="MuiButtonBase-root MuiIconButton-root PrivateSwitchBase-root-15 MuiCheckbox-root Component-root-13 PrivateSwitchBase-checked-16 Mui-checked Component-checked-14"
          >
            <span
              class="MuiIconButton-label"
            >
              <input
                checked=""
                class="PrivateSwitchBase-input-18"
                data-indeterminate="false"
                name="checkedCommitForGraphA"
                type="checkbox"
                value=""
              />
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M19 3H5c-1.11 0-2 .9-2 2v14c0 1.1.89 2 2 2h14c1.11 0 2-.9 2-2V5c0-1.1-.89-2-2-2zm-9 14l-5-5 1.41-1.41L10 14.17l7.59-7.59L19 8l-9 9z"
                />
              </svg>
            </span>
            <span
              class="MuiTouchRipple-root"
            />
          </span>
          <span
            class="MuiTypography-root MuiFormControlLabel-label MuiTypography-body1"
          >
            Commits
          </span>
        </label>
        <label
          class="MuiFormControlLabel-root"
        >
          <span
            aria-disabled="false"
            class="MuiButtonBase-root MuiIconButton-root PrivateSwitchBase-root-15 MuiCheckbox-root Component-root-19 PrivateSwitchBase-checked-16 Mui-checked Component-checked-20"
          >
            <span
              class="MuiIconButton-label"
            >
              <input
                checked=""
                class="PrivateSwitchBase-input-18"
                data-indeterminate="false"
                name="checkedMergeRequestForGraphA"
                type="checkbox"
                value=""
              />
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M19 3H5c-1.11 0-2 .9-2 2v14c0 1.1.89 2 2 2h14c1.11 0 2-.9 2-2V5c0-1.1-.89-2-2-2zm-9 14l-5-5 1.41-1.41L10 14.17l7.59-7.59L19 8l-9 9z"
                />
              </svg>
            </span>
            <span
              class="MuiTouchRipple-root"
            />
          </span>
          <span
            class="MuiTypography-root MuiFormControlLabel-label MuiTypography-body1"
          >
            Merge Requests
          </span>
        </label>
      </div>
    </div>
    <div
      class="makeStyles-container3-10"
    >
      <div
        class="makeStyles-graphContainer-11"
      >
        <p>
           (graph placeholder)            
        </p>
      </div>
      <div
        class="MuiFormGroup-root"
      >
        <label
          class="MuiFormControlLabel-root"
        >
          <span
            aria-disabled="false"
            class="MuiButtonBase-root MuiIconButton-root PrivateSwitchBase-root-15 MuiCheckbox-root Component-root-13 PrivateSwitchBase-checked-16 Mui-checked Component-checked-14"
          >
            <span
              class="MuiIconButton-label"
            >
              <input
                checked=""
                class="PrivateSwitchBase-input-18"
                data-indeterminate="false"
                name="checkedCommitForGraphB"
                type="checkbox"
                value=""
              />
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M19 3H5c-1.11 0-2 .9-2 2v14c0 1.1.89 2 2 2h14c1.11 0 2-.9 2-2V5c0-1.1-.89-2-2-2zm-9 14l-5-5 1.41-1.41L10 14.17l7.59-7.59L19 8l-9 9z"
                />
              </svg>
            </span>
            <span
              class="MuiTouchRipple-root"
            />
          </span>
          <span
            class="MuiTypography-root MuiFormControlLabel-label MuiTypography-body1"
          >
            Commits
          </span>
        </label>
        <label
          class="MuiFormControlLabel-root"
        >
          <span
            aria-disabled="false"
            class="MuiButtonBase-root MuiIconButton-root PrivateSwitchBase-root-15 MuiCheckbox-root Component-root-19 PrivateSwitchBase-checked-16 Mui-checked Component-checked-20"
          >
            <span
              class="MuiIconButton-label"
            >
              <input
                checked=""
                class="PrivateSwitchBase-input-18"
                data-indeterminate="false"
                name="checkedMergeRequestForGraphB"
                type="checkbox"
                value=""
              />
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M19 3H5c-1.11 0-2 .9-2 2v14c0 1.1.89 2 2 2h14c1.11 0 2-.9 2-2V5c0-1.1-.89-2-2-2zm-9 14l-5-5 1.41-1.41L10 14.17l7.59-7.59L19 8l-9 9z"
                />
              </svg>
            </span>
            <span
              class="MuiTouchRipple-root"
            />
          </span>
          <span
            class="MuiTypography-root MuiFormControlLabel-label MuiTypography-body1"
          >
            Merge Requests
          </span>
        </label>
      </div>
    </div>
  </div>
</div>
`;
